File: tests/page-object-model/basepage.ts
import { Locator, Page } from "@playwright/test";

export abstract class BasePage {

    readonly _popupCloseLocator: Locator;

    constructor(readonly page: Page) {
        this._popupCloseLocator = this.page.locator('"Click here to hide popup"');
    }

    protected async navigateTo(path?: string): Promise<void> {
        if (path)
            await this.page.goto(path);
        else
            await this.page.goto('/');

        await this.handleRandomPopup();
    }

    private async handleRandomPopup() {
        await this._popupCloseLocator.click();
    }
}

File: PlaywrightDotNet/PlaywrightDotNet/PageObjectModel/Pages/BasePage.cs
using PlaywrightDotNet.PageObjectModel.SharedComponents;

namespace PlaywrightDotNet.PageObjectModel.Pages;

public interface IBasePage
{
    Task NavigateToAsync();
    INavBarComponent Navbar { get; }
}

public abstract class BasePage : IBasePage
{
    public IPage Page { get; }

    private ILocator PopupCloseLocator => Page.Locator("'Click here to hide popup'");
    protected string PagePath { get; init; }
    public INavBarComponent Navbar { get; }

    protected BasePage(IPage page)
    {
        Page = page;
        Navbar = new NavBarComponent(page);
    }

    public virtual async Task NavigateToAsync()
    {
        await Page.GotoAsync($"http://localhost:4200/{PagePath}");
        await HandleRandomPopup();
    }

    private async Task HandleRandomPopup()
    {
        var popupOpen = await PopupCloseLocator.WaitForVisibleWithoutThrowing();

        if (popupOpen)
            await PopupCloseLocator.ClickAsync();
    }
}